### PR TITLE
Add note about nested parameters parsing change in rack 3

### PR DESCRIPTION
**Pull request by jarthod**
Following the discussion in https://github.com/rack/rack/issues/2128 about nested parameters parsing change in rack 3, here is a suggestion PR to mention it in the upgrade guide.

### CODE REVIEW

File: UPGRADE-GUIDE.md [hunk 1]
```diff
@@ -186,6 +186,31 @@ Previously `.rewind` was called after consuming form and multipart data. Use
 `Rack::RewindableInput::Middleware` to make the body rewindable, and call
 `.rewind` explicitly to match this behavior.
 
+### Nested query parsing does not accept the `a[b[c]]=x` syntax any more
+
+Despite being non-standard, this syntax is sometimes used by Rails
+applications because Rack 2 was (involuntarily) parsing it correctly.
+
+This syntax is no longer supported by Rack 3 and will lead to
+incorrect results:
```
Comment (jeremyevans) [inline UPGRADE-GUIDE.md:195]:
```suggestion
This invalid syntax was never officially supported.  However, some libraries
and applications used it anyway.  Due to implementation details, Rack 2
ended up parsing it the same as the correct syntax. The implementation
was changed in Rack 3, and this invalid syntax is no longer parsed the same
way as the correct syntax:
```

File: UPGRADE-GUIDE.md [hunk 2]
```diff
@@ -186,6 +186,31 @@ Previously `.rewind` was called after consuming form and multipart data. Use
 `Rack::RewindableInput::Middleware` to make the body rewindable, and call
 `.rewind` explicitly to match this behavior.
 
+### Nested query parsing does not accept the `a[b[c]]=x` syntax any more
```
Comment (jeremyevans) [inline UPGRADE-GUIDE.md:189]:
```suggestion
### Invalid nested query parsing syntax `a[b[c]]=x` no longer works the same as valid syntax `a[b][c]=x`
```

**Review by jeremyevans — CHANGES_REQUESTED** [review]
I would like this section to be more clear that the previous syntax was invalid, never officially supported, and nobody should have been using it.

**Comment by jarthod** [comment]
@jeremyevans of course, all good for me 👍

**Review by jeremyevans — APPROVED** [review]

File: UPGRADE-GUIDE.md [hunk 3]
```diff
@@ -186,6 +186,31 @@ Previously `.rewind` was called after consuming form and multipart data. Use
 `Rack::RewindableInput::Middleware` to make the body rewindable, and call
 `.rewind` explicitly to match this behavior.
 
+### Invalid nested query parsing syntax `a[b[c]]=x` no longer works the same as valid syntax `a[b][c]=x`
+
+This invalid syntax was never officially supported.  However, some libraries
+and applications used it anyway.  Due to implementation details, Rack 2
+ended up parsing it the same as the correct syntax. The implementation
+was changed in Rack 3, and this invalid syntax is no longer parsed the same
+way as the correct syntax:
```
Comment (ioquatix) [inline UPGRADE-GUIDE.md:195]: I want to keep the title a bit cleaner, but everything else looks good.

```suggestion
### Invalid nested query parsing syntax

Previously, Rack 2 was able to parse the query string `a[b[c]]=x` in the same
way as `a[b][c]=x`. This invalid syntax was never officially supported. However,
some libraries and applications used it anyway. Due to implementation details,
Rack 2 ended up parsing it the same as the correct syntax. The implementation
was changed in Rack 3, and this invalid syntax is no longer parsed the same way
as the correct syntax:
```

**Review by ioquatix — APPROVED** [review]

File: UPGRADE-GUIDE.md [hunk 4]
```diff
@@ -186,6 +186,32 @@ Previously `.rewind` was called after consuming form and multipart data. Use
...
+The correct syntax for nested parameters is `a[b][c]=x` and you'll need
+to change that in your application code to be compatible with Rack 3:
+
+```ruby
+Rack::Utils.parse_nested_query("a[b][c]=x")
+# Rack 3 => {"a"=>{"b"=>{"c"=>"x"}}} ✅
+# Rack 2 => {"a"=>{"b"=>{"c"=>"x"}}} ✅
+```
+
+See https://github.com/rack/rack/issues/2128
```
Comment (ioquatix) [inline UPGRADE-GUIDE.md:213]:
```suggestion
See <https://github.com/rack/rack/issues/2128> for more context.
```

**Pull request merged**